### PR TITLE
Update maths functions to run on Iris Xe

### PR DESCRIPTION
**Pull request by vlad-perevezentsev**
This PR modifies some of the mathematical functions to run and work correctly on Iris Xe which does not support double precision floating point type. 
Updated : `dpnp.arctan2`, `dpnp.copysign`, `dpnp.fmod`, `dpnp.hypot`, `dpnp.maximum`, `dpnp.minimum`, `dpnp.cross`
All mathematical functions will be modified in further stage so the PR is open as a draft. 

Note:
The kernel `dpnp_remainder_c` implementation uses casting to double types.
This will be fixed in another PR 

- [X] Have you provided a meaningful PR description?
- [X] Have you added a test, reproducer or referred to issue with a reproducer?
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: dpnp/backend/include/dpnp_gen_2arg_3type_tbl.hpp [hunk 1]
```diff
@@ -138,16 +138,16 @@ MACRO_2ARG_3TYPES_OP(dpnp_divide_c,
                                         std::complex<double>))
 
 MACRO_2ARG_3TYPES_OP(dpnp_fmod_c,
-                     sycl::fmod((double)input1_elem, (double)input2_elem),
-                     nullptr,
-                     std::false_type,
+                     dispatch_fmod_op(input1_elem, input2_elem),
+                     x1 % x2,
```
Comment (antonwolfy) [inline dpnp/backend/include/dpnp_gen_2arg_3type_tbl.hpp:142]: can be the `dispatch_fmod_op` used in vector implementation also?
```suggestion
                     dispatch_fmod_op(x1, x2),
```

File: dpnp/backend/kernels/dpnp_krnl_elemwise.cpp [hunk 2]
```diff
@@ -1719,25 +1730,89 @@ static void func_map_init_elemwise_2arg_3type(func_map_t &fmap)
         eft_DBL, (void *)dpnp_arctan2_c_default<double, double, double>};
 
     fmap[DPNPFuncName::DPNP_FN_ARCTAN2_EXT][eft_INT][eft_INT] = {
-        eft_DBL, (void *)dpnp_arctan2_c_ext<double, int32_t, int32_t>};
+        get_default_floating_type(),
```
Comment (antonwolfy) [inline dpnp/backend/kernels/dpnp_krnl_elemwise.cpp:1733]: can it be packed through templates like it did for DPNP_FN_DIVIDE_EXT?

File: dpnp/backend/include/dpnp_gen_2arg_3type_tbl.hpp (hunk 1 above)
Comment (vlad-perevezentsev) [reply dpnp/backend/include/dpnp_gen_2arg_3type_tbl.hpp]: Yes, sure!
Done

File: dpnp/backend/src/dpnp_fptr.hpp [hunk 3]
```diff
@@ -271,6 +271,62 @@ static constexpr DPNPFuncType get_default_floating_type()
...
+        }
+    }
+
+    // Default case
+    if constexpr (has_fp64::value) {
+        return DPNPFuncType::DPNP_FT_DOUBLE;
+    }
+    else {
+        return DPNPFuncType::DPNP_FT_FLOAT;
+    }
```
Comment (antonwolfy) [inline dpnp/backend/src/dpnp_fptr.hpp:327]:
```suggestion
    return get_default_floating_type<has_fp64>();
```

File: dpnp/backend/src/dpnp_fptr.hpp [hunk 4]
```diff
@@ -271,6 +271,62 @@ static constexpr DPNPFuncType get_default_floating_type()
...
+        else if constexpr (shortes_type == DPNPFuncType::DPNP_FT_INT ||
+                           shortes_type == DPNPFuncType::DPNP_FT_LONG)
+        {
+            // Check if has_fp64
+            if constexpr (has_fp64::value) {
+                return DPNPFuncType::DPNP_FT_DOUBLE;
+            }
+            else {
+                return DPNPFuncType::DPNP_FT_FLOAT;
+            }
```
Comment (antonwolfy) [inline dpnp/backend/src/dpnp_fptr.hpp:317]: I guess it might be removed since the same is done at line#321

File: dpnp/backend/src/dpnp_fptr.hpp [hunk 5]
```diff
@@ -271,6 +271,62 @@ static constexpr DPNPFuncType get_default_floating_type()
...
 
+/**
+ * A template function that determines the resulting floating-point type
+ * based on the value of the template parameter has_fp64.
+ */
+template <DPNPFuncType FT1,
+          DPNPFuncType FT2,
+          typename has_fp64 = std::true_type,
+          typename keep_int = std::false_type>
+static constexpr DPNPFuncType get_res_type()
```
Comment (antonwolfy) [inline dpnp/backend/src/dpnp_fptr.hpp:282]:
```suggestion
static constexpr DPNPFuncType get_floating_res_type()
```

File: tests/test_mathematical.py [hunk 6]
```diff
@@ -162,6 +163,16 @@ def test_divide(self, dtype, lhs, rhs):
         "dtype", get_all_dtypes(no_bool=True, no_complex=True)
     )
     def test_fmod(self, dtype, lhs, rhs):
+        if dtype == None and rhs == 0.3 and not has_support_aspect64():
+            """
+            Due to some reason NumPy behaves incorrectly, when:
```
Comment (antonwolfy) [inline tests/test_mathematical.py:168]:
```suggestion
            Due to accuracy reason NumPy behaves differently, when:
```

File: tests/test_mathematical.py [hunk 7]
```diff
@@ -162,6 +163,16 @@ def test_divide(self, dtype, lhs, rhs):
...
+                >>> numpy.fmod(numpy.array([3.9], dtype=numpy.float32), 0.3)
+                array([0.29999995], dtype=float32)
+            while numpy with float64 returns something around zero which is expected:
```
Comment (antonwolfy) [inline tests/test_mathematical.py:171]:
```suggestion
            while numpy with float64 returns something around zero which is aligned with dpnp:
```

File: tests/test_mathematical.py [hunk 8]
```diff
@@ -768,9 +782,9 @@ def test_inplace_strided_out(self, dtype):
         "shape", [(0,), (15,), (2, 2)], ids=["(0,)", "(15, )", "(2,2)"]
     )
     def test_invalid_shape(self, shape):
-        dp_array1 = dpnp.arange(10, dtype=dpnp.float64)
-        dp_array2 = dpnp.arange(5, 15, dtype=dpnp.float64)
-        dp_out = dpnp.empty(shape, dtype=dpnp.float64)
+        dp_array1 = dpnp.arange(10, dtype=dpnp.float32)
```
Comment (antonwolfy) [inline tests/test_mathematical.py:785]: I believe we are free to remove dtype argument here at all

**Review by antonwolfy — APPROVED** [review]
LGTM! Thank you @vlad-perevezentsev

**Pull request merged**